packages/protocol/contracts/common/UUPSUpgradeableAlt.sol

```diff
@@ -0,0 +1,114 @@
+// SPDX-License-Identifier: MIT
+//  _____     _ _         _         _
+// |_   _|_ _(_) |_____  | |   __ _| |__ ___
+//   | |/ _` | | / / _ \ | |__/ _` | '_ (_-<
+//   |_|\__,_|_|_\_\___/ |____\__,_|_.__/__/
+//
+//   Email: security@taiko.xyz
+//   Website: https://taiko.xyz
+//   GitHub: https://github.com/taikoxyz
+//   Discord: https://discord.gg/taikoxyz
+//   Twitter: https://twitter.com/taikoxyz
+//   Blog: https://mirror.xyz/labs.taiko.eth
+//   Youtube: https://www.youtube.com/@taikoxyz
+
+pragma solidity 0.8.20;
+
+import "lib/openzeppelin-contracts/contracts/interfaces/draft-IERC1822.sol";
+import "lib/openzeppelin-contracts/contracts/proxy/ERC1967/ERC1967Upgrade.sol";
+
+/**
+ * @dev An upgradeability mechanism designed for UUPS proxies.
+ * This is an alternative impl based on
+ * lib/openzeppelin-contracts/contracts/proxy/utils/UUPSUpgradeable.sol
+ * without the immutable variable so we can pre-deploy these contracts on L2.
+ */
+abstract contract UUPSUpgradeableAlt is IERC1822Proxiable, ERC1967Upgrade {
+    uint256[50] private __gap;
+
+    error DISALLOWED_WHEN_NOT_DELEGAATED();
+    error DISALLOWED_WHEN_IS_DELEGATED();
+    /**
+     * @dev Check that the execution is being performed through a delegatecall call and that the
+     * execution context is a proxy contract with an implementation (as defined in ERC1967).
+     */
+
+    modifier onlyProxy() {
```

Comment:
For these two modifiers, please compare the current implementations with the original ones in `lib/openzeppelin-contracts/contracts/proxy/utils/UUPSUpgradeable.sol`

Reply:
@Brechtpd Please also help me review the code.

Reply:
I think this check it less strong than the original one because, without `_self`,  it's impossible here to check that the correct implementation contract is being called without having access to the "real" `address(this)` of the implementation contract. So we can only check that something is delegated, but not that it is delegated to the correct implementation contract, which is what the original `onlyProxy` does check.

Reply:
I also prefer to substitute the address in the compiled bytecode to "initialize" immutable variables correctly. Lets wait for @davidtaikocha on that.

Reply:
is there a strong reason for the switch to this proxy versus the pre existing TtansparentUpgradableProxy we had before which I believe doesn't have this issue (since we were able to upgrade contacts before?)

afaik, main feature of this proxy is you can eventually disable upgradability.

is it better to patch this, or use an out of the box proxy solution that's worked for us in the past?